src/formatter.rs

```diff
@@ -15,6 +15,10 @@ fn highlight_command<'a>(
     config: &Config,
     parts: &mut Vec<ANSIString<'a>>,
 ) {
+    if command.trim().is_empty() {
+        parts.push(config.style.example_code.paint(example_code));
+        return;
+    }
```

Comment:
Your are right. This part and the one after the while-loop is basically the same. On the other hand, I like to return early and avoid high nesting levels. Wrapping the loop into an if-block, would already cause a nesting level of 4 in this method. Static analysis tools typically don't like this for a good reason. Sometimes it is just fine to "duplicate" a little bit if it increases readability and reduces cognitive load.

Reply:
You make a good point. However, I am still against this change, because I think that this function is already hard to read and this would make it harder. I got a bit hung up on what this function does and looked up the original proposal: https://github.com/dbrgn/tealdeer/commit/66b8a3dacacad4ba8bf27db2c9fb46b4275aa8b4#diff-cffebbb289ff3011edae459ca9e6843092b4b3c45a757f4964b8f633ca669c1eR12-R27

To me, this seems like it just wants to output the line (`example_code` should in my eyes just be called `line` or so) and highlights everything as example code, except the command name, which has special styling. I think, that this can be written more concise and clearly like this:

```rust
    #[derive(Clone, Copy)]
    enum HighlightingType {
        CommandName,
        OtherCode(&str),
    }
    line.split(command)
        .map(HighlightingType::OtherCode)
        .intersperse(HighlightingType::CommandName)
        .for_each(|token| match token {
            HighlightingType::CommandName => parts.push(config.style.command_name.paint(command)),
            HighlightingType::OtherCode(s) => parts.push(config.style.example_code.paint(s)),
        });
```

This is _a lot_ clearer to me than the linked code. This has some problems though. Firstly, this uses `Iterator::intersperse` which is a nightly feature right now. There is something similar in the `itertools` crate though which can be used with stable. Secondly (and more imporantly), this does not handle the added checks from #157. So `split` might not cut it for us here.

Looking at the methods available to us with `&str`, I think we can use [`matches_indicies`](https://doc.rust-lang.org/std/primitive.str.html#method.match_indices) to only the starts of the `command` substrings and perform similar checks there.

---

Now, I have spiraled from the logic of this PR (again). You see how complicated this function is for what it does. If you find a way to refactor this similar to the above but without chaning the behavior, that would be great, otherwise I might open a PR to replace the code myself. Maybe you can find an approach that even handles the case that `command` is whitespace-only without explicit checking. I am not sure how the searching semantics are for say an empty string. If not, I guess we can incorporate the check into the function too (the current state is no single ones fault after all / not even wrong), but I think we should give refactoring a try while we're at it! 😄

Reply:
I agree with you that this code is not easily understandable anymore. On the other hand, the logic behind it is not as simple as it might seem. After some playing around, I was not able to simplify it significantly with standard methods. 

What I did now is trying the parser combinator library [nom](https://github.com/Geal/nom). You can find an implementation using it [here](https://github.com/SimplyDanny/tealdeer/commit/a6ebab63a5635a26e15b9c50423124ebb0f70103). The code is much simpler. It even fixes another highlighting issue with having the command name indeed preceded by a space but not followed by a space, which would require to make the existing code still more complex in order to fix it. The command name itself can be empty without a separate check. As a drawback I see that it is yet another dependency and it might potentially have a negative effect on performance. However, this has yet to be evaluated.

What do you think?

Reply:
In general I'd prefer not to bring in an additional dependency, unless it adds a lot of benefits. Cleaner code _is_ a benefit, but I'm still unsure whether it is worth it.

> otherwise I might open a PR to replace the code myself

@niklasmohrin maybe you want to give it a try? If you come up with an approach that does not require nom but still makes the code easier to understand, that would be cool. (It seems you had some ideas...)

(Regarding nightly: Tealdeer should be able to build on stable. Using itertools temporarily until it's stabilized would be fine though.)

Reply:
I could try to implement what I thought out before, it may take a couple of days though since I am rather occupied currently. About nightly: Agree.

I feel like `nom` is a bit heavy for the simple use case we have and to be honest, I had a hard time reading the parser logic. It does handle the variable highlighting more clearly though. Still, I don't think that it suits this problem well enough to approve it right now and I would only come back to this idea if the rest doesn't work out :man_shrugging:

Reply:
See #189 for implementation

Reply:
FYI, #189 is now merged.